In [ ]:
import math
import numpy as np

In [ ]:
def forward_diff(f_1,f_0,h):
    f_prime = (f_1-f_0)/h
    return f_prime

def back_diff(f_1,f_0,h):
    f_prime = (f_0-f_1)/h
    return f_prime
     
     

# $ \S 4.1 $

## Exercise 1 

Use the forward-difference formulas and backward-difference formulas to determine each missing entry in the following tables.

### Part (a)


In [ ]:
x=[0.5,0.6,0.7]
f_x = [0.4794,0.5646,0.6442]

part_a_list = []
part_a_list.append(forward_diff(f_x[1],f_x[0],0.1))
part_a_list.append(forward_diff(f_x[2],f_x[1],0.1))
part_a_list.append(back_diff(f_x[1],f_x[2],0.1))

part_a = np.round(part_a_list,4)

print(part_a)

[0.852 0.796 0.796]


### part (b)

$x=[0.0,0.2,0.4]$, $f(x) = [0.00000,0.74140,1.3718]$

In [ ]:
x=[0.0,0.2,0.4]
f_x = [0.00000,0.74140,1.3718]

part_b_list = []
part_b_list.append(forward_diff(f_x[1],f_x[0],0.2))
part_b_list.append(forward_diff(f_x[2],f_x[1],0.2))
part_b_list.append(back_diff(f_x[1],f_x[2],0.2))

part_b = np.round(part_b_list,4)

print(part_b)


[3.707 3.152 3.152]


$f'(0.0)\approx 3.707,\,  f'(0.2)\approx 3.152\, f'(0.4)\approx 3.152$

## Exercise 3

The data in Exercise 1 were taken from the following functions. Compute the actual errors in Exercise 1 , and find error bounds using the error formulas.

### part (a)

$f(x)=\sin x$

$f'(x) = \cos x$

In [ ]:
x=[0.5,0.6,0.7]
f_x = [0.4794,0.5646,0.6442]
for i in range(len(x)):
    print(f'Actual error at {x[i]} : {np.abs(np.cos(x[i])-part_a[i]):.5f}')
    
    

Actual error at 0.5 : 0.02558
Actual error at 0.6 : 0.02934
Actual error at 0.7 : 0.03116


Error bound:

We have $f''(x) = - \sin(x)$

$$
\left|\frac{h}{2} f^{(2)}(\xi)\right| \leq 0.05\left(\max _{\xi \in[a a+h]}|-\sin (\xi)|\right)
$$

$\sin(x)$ is increasing on $[0.5,0.6]$ and $[0.6,0.7]$

The bound associated with $x=0.5$ and $x=0.6$ will remain the same thus we take $\xi = 0.6$

$0.05\left(\max _{\xi \in[a a+h]}|-\sin (\xi)|\right) = 0.05\times0.5646=0.02823$

For $x =0.7$ we take $\xi=0.7$

$0.05\left(\max _{\xi \in[a a+h]}|-\sin (\xi)|\right) = 0.05\times0.6442=0.03221$

### part (b)

b. $f(x)=e^{x}-2 x^{2}+3 x-1$

$f'(x) = e^{x} -4x+3$

In [224]:
x=[0.0,0.2,0.4]
for i in range(len(x)):
    print(f'Actual error at {x[i]} : {(np.exp(x[i])-4*x[i]+3)-part_b[i]:.5f}')
    
    

    

Actual error at 0.0 : 0.29300
Actual error at 0.2 : 0.26940
Actual error at 0.4 : -0.26018


Error bound:

We have $f''(x) =  e^x-4$

$$
\left|\frac{h}{2} f^{(2)}(\xi)\right| \leq 0.05\left(\max _{\xi \in[a a+h]}|e^x-4|\right)
$$

$e^x$ is increasing on $[0.0,0.2]$ and $[0.2,0.4]$


For $x =0.0$ we take $\xi=0.0$ to maximize 

$0.1\left(\max _{\xi \in[a a+h]}|e^x-4|\right) = 0.1\times 3=0.3$

The bound associated with $x=0.2$ and $x=0.4$ will remain the same thus we take $\xi = 0.2$ to maximize

$0.1\left(\max _{\xi \in[a a+h]}|e^x-4|\right) = 0.1\times0.5646=0.02823$

For $x =0.7$ we take $\xi=0.7$

$0.1\left(\max _{\xi \in[a a+h]}|e^x-4|\right) = 0.1\times2.77859=0.277859$

## Exercise 5

Use the most accurate three-point formula to determine each missing entry in the following tables.

### part (a)
$$
\begin{aligned}
&\begin{array}{c|c|c}
x & f(x) & f^{\prime}(x) \\
\hline 1.1 & 9.025013 & \\
1.2 & 11.02318 & \\
1.3 & 13.46374 & \\
1.4 & 16.44465 &
\end{array}
\end{aligned}
$$

In [225]:
def three_endpoint(f_x0,f_x1,f_x2,h):
    f_prime = (1/(2*h))*(-3*(f_x0)+4*f_x1 -f_x2)
    return f_prime

In [226]:
three_endpoint(9.025013,11.02318,13.46374,0.1)

def three_midpoint(f_x0,f_x1,h):
    f_prime = (1/(2*h))*((f_x1)-(f_x0))
    return f_prime

In [227]:
num_5_pa = []
num_5_pa.append(three_endpoint(9.025013,11.02318,13.46374,0.1))
num_5_pa.append(three_midpoint(9.025013,13.46374,0.1))
num_5_pa.append(three_midpoint(11.02318,16.44465,0.1))
num_5_pa.append(three_endpoint(16.44465,13.46374 ,11.02318,-0.1))

num_5_pa = np.round(num_5_pa,5)

print(num_5_pa)

[17.76971 22.19364 27.10735 32.51085]


$$
\begin{aligned}
&\begin{array}{c|c|c}
x & f(x) & f^{\prime}(x) \\
\hline 1.1 & 9.025013 & 17.76971\\
1.2 & 11.02318 & 22.19364\\
1.3 & 13.46374 & 27.10735 \\
1.4 & 16.44465 & 32.51085
\end{array}
\end{aligned}
$$

## part (d)

$$
\begin{aligned}
&\begin{array}{c|c|c}
x & f(x) & f^{\prime}(x) \\
\hline 2.0 & 3.6887983 & \\
2.1 & 3.6905701 & \\
2.2 & 3.6688192 & \\
2.3 & 3.6245909 &
\end{array}
\end{aligned}
$$

In [228]:
num_5_pb = []

num_5_pb.append(three_endpoint(3.6887983,3.6905701,3.6688192,0.1))
num_5_pb.append(three_midpoint(3.6887983,3.6688192,0.1))
num_5_pb.append(three_midpoint(3.6905701,3.6245909,0.1))
num_5_pb.append(three_endpoint(3.6245909,3.6688192,3.6905701,-0.1))

num_5_pb = np.round(num_5_pb,7)

print(num_5_pb)

[ 0.1353315 -0.0998955 -0.329896  -0.55467  ]


$$
\begin{aligned}
&\begin{array}{c|c|c}
x & f(x) & f^{\prime}(x) \\
\hline 2.0 & 3.6887983 & 0.13533\\
2.1 & 3.6905701 & -0.0999\\
2.2 & 3.6688192 &-0.3299 \\
2.3 & 3.6245909 &-0.55467
\end{array}
\end{aligned}
$$

## Exercise 7

The data in Exercise 5 were taken from the following functions. Compute the actual errors in Exercise 5 , and find error bounds using the error formulas.

### Part (a)

$f(x) = e^{2x}$

$f'(x) = 2e^{2x}$, $f''(x) = 4e^{2x}$, $f^{(3)}=8e^{2x}$

In [230]:
x=[1.1,1.2,1.3,1.4]
for i in range(len(x)):
    print(f'Actual error at {x[i]} : {np.abs((2*np.exp(2*x[i]))-num_5_pa[i]):.5f}')

Actual error at 1.1 : 0.28032
Actual error at 1.2 : 0.14729
Actual error at 1.3 : 0.17987
Actual error at 1.4 : 0.37844


$f'(1.1)\approx 0.28032,\,  f'(1.2)\approx 0.14729,\, f'(1.3)\approx 0.17987,\,f'(1.4)\approx 0.37844$

For endpoints we have error $
\frac{h^{2}}{3} f^{(3)}(\xi)$ the bound $\dfrac{Mh}{3}$ where $M$ is a bound on $|f^{(3)}|$ and $\xi$ is in between $x=1.1$ and $x=1.3$. $8e^{2x}$ is an increasing function, hence we take $1.3$ to maximize. We have


for $x=1.1$ $$\dfrac{Mh}{3}=\frac{8e^{2\left(1.3\right)}\left(0.1\right)^{2}}{3}=0.359033
$$

For midpoints we have error $
\frac{h^{2}}{6} f^{(3)}(\xi)$, the bound $\dfrac{Mh}{6}$ where $M$ is a bound on $|f^{(3)}|$ and $\xi$ is in between $x=1.2$ and $x=1.3$. $8e^{2x}$ is an increasing function, hence we take $1.3$ to maximize. We have

for $x=1.2$ $$\dfrac{Mh}{6}=\frac{8e^{2\left(1.3\right)}\left(0.1\right)^{2}}{6}=0.17951
$$

for $x=1.3$ we take $x=1.4$ to maximize $$\dfrac{Mh}{6}=\frac{8e^{2\left(1.4\right)}\left(0.1\right)^{2}}{6}=0.219262
$$

For the last endpoint we take the endpoint bound but instead with $x$ in $x_0$ and $x_0-2h$ with $-h$, namely $[1.2,1.4]$. Thus,

for $x=1.4$ we take $x=1.4$ to maximize $$\dfrac{Mh}{3}=\frac{8e^{2\left(1.4\right)}\left(-0.1\right)^{2}}{3}=0.4385239$$


### part (d) 

$f(x) = 2(\ln x)^2 + 3\sin x$

$f'(x)= \dfrac{4(\ln(x))}{x}+3\cos x$

$
f^{(3)}(x)=\frac{-12}{x^{3}}+\frac{8 \ln x}{x^{3}}-3 \cos x
$

In [239]:
x=[2.0,2.1,2.2,2.3]
for i in range(len(x)):
    print(f'Actual error at {x[i]} : {np.abs((4*np.log(x[i])/x[i])+3*np.cos(x[i])-num_5_pb[i]):.7f}')

Actual error at 2.0 : 0.0025224
Actual error at 2.1 : 0.0014288
Actual error at 2.2 : 0.0020485
Actual error at 2.3 : 0.0043795


$f'(2.0)\approx 0.0025224,\,  f'(2.1)\approx 0.0014288,\, f'(2.2)\approx 0.0020485,\,f'(2.3)\approx 0.0043795$

For endpoints we have error $
\frac{h^{2}}{3} f^{(3)}(\xi)$ the bound $\dfrac{Mh}{3}$ where $M$ is a bound on $|f^{(3)}|$ and $\xi$ is in between $x=2.0$ and $x=2.2$. $f^{(3)}(x)=\frac{-12}{x^{3}}+\frac{8 \ln x}{x^{3}}-3 \cos x$ is an increasing function on $[2.0.2.3]$, hence we take $2.2$ to maximize. We have


for $x=2.0$ 
$$\dfrac{Mh}{3}=\left(\frac{\left(\frac{-12}{\left(2.2\right)^{3}}+\frac{8\ln\left(2.2\right)}{\left(2.2\right)^{3}}-3\cos\left(2.2\right)\right)}{3}\right)\cdot\left(0.1\right)^{2}=0.00410303611712$$

For $x=2.1$ we take $x=2.2$ to maximize using midpoint bound:

$$\dfrac{Mh}{6}=\left(\frac{\left(\frac{-12}{\left(2.2\right)^{3}}+\frac{8\ln\left(2.2\right)}{\left(2.2\right)^{3}}-3\cos\left(2.2\right)\right)}{6}\right)\cdot\left(0.1\right)^{2}=0.002051518$$

For $x=2.2$ we take $x=2.3$ to maximize using midpoint bound:

$$\dfrac{Mh}{6}=\left(\frac{\left(\frac{-12}{\left(2.3\right)^{3}}+\frac{8\ln\left(2.3\right)}{\left(2.3\right)^{3}}-3\cos\left(2.3\right)\right)}{6}\right)\cdot\left(0.1\right)^{2}=0.0026003$$

For endpoint $x=2.3$ we have $x\in [2.1,2.3]$, thus $x=2.3$ to maximize

for $x=2.3$ 
$$\dfrac{Mh}{3}=\left(\frac{\left(\frac{-12}{\left(2.3\right)^{3}}+\frac{8\ln\left(2.3\right)}{\left(2.3\right)^{3}}-3\cos\left(2.3\right)\right)}{3}\right)\cdot\left(0.1\right)^{2}=0.00520068$$

22. In a circuit with impressed voltage $\mathcal{E}(t)$ and inductance $L$, Kirchhoff's first law gives the relationship
$$
\mathcal{E}(t)=L \frac{d i}{d t}+R i
$$
where $R$ is the resistance in the circuit and $i$ is the current. Suppose we measure the current for several values of $t$ and obtain
$$
\begin{array}{l|l|l|l|l|l}
t & 1.00 & 1.01 & 1.02 & 1.03 & 1.04 \\
\hline i & 3.10 & 3.12 & 3.14 & 3.18 & 3.24
\end{array}
$$
where $t$ is measured in seconds, $i$ is in amperes, the inductance $L$ is a constant $0.98$ henries, and the resistance is $0.142 \mathrm{ohms}$. Approximate the voltage $\mathcal{E}(t)$ when $t=1.00,1.01,1.02,1.03$, and $1.04$.

We will use 3-point formula:
$R = 0.142, L=0.98$

$h = 0.01$

for $t = 1.00$ we use enpoint formula

In [252]:
t_00=np.round(three_endpoint(3.10,3.12,3.14,0.01),5)

for $t=1.01,1.02,1.03$ we use midpoint formula

In [248]:
t_01=np.round(three_midpoint(3.10,3.14,.01))
t_02=np.round(three_midpoint(3.12,3.18,.01))
t_03=np.round(three_midpoint(3.14,3.24,.01))

for $t=1.04$ we use endpoint formula with $-h$

In [255]:
t_04=np.round(three_endpoint(3.24,3.18,3.14,-0.01),5)

In [256]:
t_list = []
t_list.extend([t_00,t_01,t_02,t_03,t_04])

In [257]:
t_list

[2.0, 2.0, 3.0, 5.0, 7.0]

$\dfrac{di}{dt}=\{2.0,2.0,3.0,5.0,7.0\}\quad \text{for}\quad t=1.00,1.01,1.02,1.03,1.04$

In [263]:
L = 0.98
R = 0.142
i=[3.10,3.12,3.14,3.18,3.24]
for j in range(len(t_list)):
    print(np.round(L*t_list[j]+ R*i[j],5))

2.4002
2.40304
3.38588
5.35156
7.32008


$$
\begin{aligned}
\mathcal{E}(1.00)&=2.4002\\
\mathcal{E}(1.01)&=240304\\
\mathcal{E}(1.02)&=3.38588\\
\mathcal{E}(1.03)&=5.35156\\
\mathcal{E}(1.04)&=7.32008
\end{aligned}
$$

## Exercise 29

Consider the function
$$
e(h)=\frac{\varepsilon}{h}+\frac{h^{2}}{6} M
$$
where $M$ is a bound for the third derivative of a function. Show that $e(h)$ has a minimum at $\sqrt[3]{3 \varepsilon / M}$.

We have $e(h)=\frac{\varepsilon}{h}+\frac{h^{2}}{6} M$, where $\mathrm{M}$ is the bound for the third derivative of a function.
Taking the derivative we get $e^{\prime}(h)=-\varepsilon / h^{2}+h M / 3$, If $e^{\prime}(h)=0$ we have $h=\sqrt[3]{3 \varepsilon / M}$. Now, $e^{\prime}(h)<0$ if $h<\sqrt[3]{3 \varepsilon / M}$ and $e^{\prime}(h)>0$ if $h>\sqrt[3]{3 \varepsilon / M}$
So  $e(h)$ has a minimum at $h=\sqrt[3]{3 \varepsilon / M}$.

# $\S 4.3$
## Exercise 1
Approximate the following integrals using the Trapezoidal rule.

a. $\int_{0.5}^{1} x^{4} d x$

d. $\int_{0}^{1} x^{2} e^{-x} d x$

g. $\int_{0}^{\pi / 4} x \sin x d x$


In [68]:
def trap_rule(a,b):
    return (b-a)*(f(a)+f(b))/2

### part (a)

In [69]:
def f(x):
    return x**4

np.round(trap_rule(0.5,1),5)

0.26562

### part (d)

In [ ]:
def func(x):
    return x**2*np.exp(-x)
trap_rule(0,1)

0.18393972058572117

### part(g)

In [28]:
def func(x):
    return x*np.sin(x)
trap_rule(0,np.pi/4)

0.21808950623871498

## Exercise 3

Find a bound for the error in Exercise 1 using the error formula and compare this to the actual error.

### part (d) 

$\begin{aligned} \int_{0}^{1} x^{2} e^{-x} d x &=-e^{-x}\left(x^{2}+2 x+2\right)_{0}^{1} \\ &=0.160602794 \end{aligned}$

True error from Exercise 1 (d) we have 

$|0.160602794-0.1839397|\approx 0.0233369$

$f''(x)=e^{-x}(2-4x+x^{2})$

We have error $\dfrac{h^{3}}{12} f^{\prime \prime}(\xi), \quad \text { where } \quad x_{0}<\xi<x_{1}$

 $|f^{\prime \prime}(\xi)|$ is maximized at $x=0$ 
 
 $\dfrac{h^{3}}{12} |f^{\prime \prime}(\xi)|\leq 2\cdot\dfrac{1^3}{12}\approx .166666666667$



### part (g)


$$\begin{aligned}
\int_{0}^{\frac{\pi}{4}} x \sin x d x &=[-x \cos x+\sin x]_{0}^{\frac{\pi}{4}} \\
&=0.152027958
\end{aligned}$$

From Exercise (1) we have the Actual Error

$|0.152027958-0.218089|\approx 0.066061$


$f''(x) = 2\cos(x)-x\sin(x)$ 

$|f''(x)|$ is maximized at at $x=0$ 

 $\dfrac{h^{3}}{12} |f^{\prime \prime}(\xi)|\leq 2\cdot\dfrac{(\pi/4)^3}{12}\approx 0.0807455$

# Exercise 5

Repeat Exercise 1 using Simpson's rule.

a. $\int_{0.5}^{1} x^{4} d x$

d. $\int_{0}^{1} x^{2} e^{-x} d x$




In [9]:
def simpson(a,b,n):

    h = (b-a) / n
    
    # Finding sum 
    integ = f(a) + f(b)
    
    for i in range(1,n):
        k = a + i*h
        
        if i%2 == 0:
            integ = integ + 2 * f(k)
        else:
            integ = integ + 4 * f(k)
    
    # Finding final integration value
    integ = integ * h/3
    
    return integ

def f(x): return x**4

simpson(0.5, 1.0, 2)


0.19401041666666666

### Part (a)

In [10]:
np.round(simpson(0.5, 1, 2),7)

0.1940104

Part (b)

In [57]:
def f(x): return x**2*np.exp(-x)

np.round(simpson(0, 1, 2),7)

0.1624017

### Exercise 15. 

Approximate the following integrals using formulas (4.25) through (4.32). Are the accuracies of the approximations consistent with the error formulas?

a. $\int_{0}^{0.1} \sqrt{1+x}$

d. $\int_{0}^{1} x^{1 / 3} d x$

In [82]:
def simpson38(a,b,n):
    h = (b-a) / n
    integ = f(a) + f(b)
    integ = integ + 3 * f(a+h)+3*f(b-h)
    integ = integ * (3*h)/8
    return integ


def simpson4(a,b,n):
    h = (b-a) / n
    integ = 7*f(a) + 7*f(b)
    integ = integ + 32 * f(a+h)+12*f(a+2*h)+32*f(a+3*h)
    integ = integ * (2*h)/45
    return integ


def open0(a,b):
    x1 = (b+a) /2
    h= (b-a)/2
    return 2*h*f(x1)

def open1(a,b):
    x1 = (b+a) /(3)
    h= (b-a)/3
    x2 = x1+h
    return (3*h)/2*(f(x1)+f(x2))

def open2(a,b):
    x1 = (b+a) /(4)
    h= (b-a)/4
    x2 = x1+h
    x3=x2+h
    return (4*h)/3*(2*f(x1)-f(x2)+2*f(x3))


def open3(a,b):
    x1 = (b+a) /(5)
    h= (b-a)/5
    x2 = x1+h; x3=x2+h; x4=x3+h
    return (5*h)/24*(11*f(x1)+f(x2)+f(x3)+11*f(x4))


### part (a)

In [113]:
def f(x): return np.sqrt(1+x)

parta_15closed=[]
parta_15open=[]
parta_15closed.append(trap_rule(0,0.1))
parta_15closed.append(simpson(0,0.1,2))
parta_15closed.append(simpson38(0, 0.1, 3))
parta_15closed.append(simpson4(0, 0.1, 4))
parta_15open.append(open0(0,0.1))
parta_15open.append(open1(0,0.1))
parta_15open.append(open2(0,0.1))
parta_15open.append(open3(0,0.1))
l1a= np.array(parta_15closed)
l2a=np.array(parta_15open)
print("Evaluation of Formulas 4.25-4.32, respectively\n")
print(l1a,l2a)

Evaluation of Formulas 4.25-4.32, respectively

[0.10244044 0.10245982 0.10245982 0.10245982] [0.10246951 0.10246628 0.10245982 0.10245982]


True value $\int_{0}^{0.1} \sqrt{1+x}=.1024598$

Error for closed and open formulas:

In [106]:
print(abs(.1024598 - l1))
print(abs(.1024598 - l2))

[1.93575915e-05 1.92425665e-08 2.07691433e-08 2.19908594e-08]
[9.70765960e-06 6.48022269e-06 2.43956157e-08 2.36623756e-08]


the accuracies of the approximations are consistent with the error formulas

In [115]:
def f(x): return x**(1/3)

partb_15closed=[]
partb_15open=[]
partb_15closed.append(trap_rule(0,1))
partb_15closed.append(simpson(0,1,2))
partb_15closed.append(simpson38(0, 1, 3))
partb_15closed.append(simpson4(0, 1, 4))
partb_15open.append(open0(0,1))
partb_15open.append(open1(0,1))
partb_15open.append(open2(0,1))
partb_15open.append(open3(0,1))
l1b= np.array(partb_15closed)
l2b=np.array(partb_15open)
print("Evaluation of Formulas 4.25-4.32, respectively\n")
print(l1b,l2b)


Evaluation of Formulas 4.25-4.32, respectively

[0.5        0.69580035 0.71260315 0.73063414] [0.79370053 0.78347087 0.76111371 0.75935723]


True value $\int_{0}^{1} \sqrt{1+x}=.75$

Error for closed and open formulas

In [116]:
print(abs(.75 - l1b))
print(abs(.75- l2b))

[0.25       0.05419965 0.03739685 0.01936586]
[0.04370053 0.03347087 0.01111371 0.00935723]


the accuracies of the approximations not consistent with the error formulas